packages/protocol/contracts/bridge/libs/LibBridgeRelease.sol

```diff
@@ -56,37 +60,99 @@ library LibBridgeRelease {
 
         bytes32 msgHash = message.hashMessage();
 
-        if (state.etherReleased[msgHash] == true) {
-            revert B_ETHER_RELEASED_ALREADY();
-        }
-
         if (
             !LibBridgeStatus.isMessageFailed(
                 resolver, msgHash, message.destChainId, proof
             )
         ) {
             revert B_MSG_NOT_FAILED();
         }
+        
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.ETH_AND_TOKEN_RELEASED
+        ){
+            // Both ether and tokens are released
+            revert B_TOKENS_RELEASED_ALREADY();
+        }
+
+        uint256 releaseAmount;
 
-        state.etherReleased[msgHash] = true;
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.NOT_RECALLED
+        ) {
+            // Release ETH first
+            state.recallStatus[msgHash] = LibBridgeData.RecallStatus.ETH_RELEASED;
 
-        uint256 releaseAmount = message.depositValue + message.callValue;
+            releaseAmount = message.depositValue + message.callValue;
 
-        if (releaseAmount > 0) {
-            address ethVault = resolver.resolve("ether_vault", true);
-            // if on Taiko
-            if (ethVault != address(0)) {
-                EtherVault(payable(ethVault)).releaseEther(
-                    message.owner, releaseAmount
-                );
-            } else {
-                // if on Ethereum
-                (bool success,) = message.owner.call{ value: releaseAmount }("");
-                if (!success) {
-                    revert B_FAILED_TRANSFER();
+            if (releaseAmount > 0) {
+                address ethVault = resolver.resolve("ether_vault", true);
+                // if on Taiko
+                if (ethVault != address(0)) {
+                    EtherVault(payable(ethVault)).releaseEther(
+                        message.owner, releaseAmount
+                    );
+                } else {
+                    // if on Ethereum
+                    (bool success,) = message.owner.call{ value: releaseAmount }("");
+                    if (!success) {
+                        revert B_FAILED_TRANSFER();
+                    }
+                }
+            }
+        }
+        //2nd stage is releasing the tokens
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.ETH_RELEASED 
+                && message.to != address(0)
+                && message.data.length != 0
```

Comment:
maybe we should not check if data.length == 0, instead, we should check if `to` implement a specific EIP165 interface with a function `recallMessage()`.

Reply:
```solidity
if(state.recallStatus[msgHash]== LibBridgeData.RecallStatus.ETH_RELEASED) {
  if (message.to == address(0) || !EIP165(message.to).supportsInterface(BridgeMessageRecallReceiver.INTERFACE) {
  state.recallStatus[msgHash] = FULLY_RECALLED;
} else {
   try(message.to.onMessageRecalled()) {
   state.recallStatus[msgHash] = FULLY_RECALLED;
} catch ()
}

}